### ISSUE TITLE
Interpolate formula fail on corner conditions

### ISSUE THREAD

**Issue by lidanqing-intel** [oneapi-src/oneDNN#918]
# Summary
Hi, I doubt there is some error in formula in https://docs.oneapi.com/versions/latest/onednn/dev_guide_resampling.html

I implemented an interface calling oneDNN resampling kernel in PaddlePaddle. When working on unittest, I have some problems.
If I use formula in the API to make expected results
![image](https://user-images.githubusercontent.com/47151829/101854143-390dbc00-3b61-11eb-903d-fe803937add6.png)
All unittest failed.

If I use following formula
`src(n, ic, ih, iw) = src(n, ic, min(max(ih, 0), IH-1), min(max(iw, 0), IW-1))`
some unit tests passed, but some still fail.

Can you please suggest if following calculation is correct? Thank you

# URLs
oneDNN Interpolate doc: https://docs.oneapi.com/versions/latest/onednn/dev_guide_resampling.html

# Additional details
My unittest reference results are calculated as follows:
```
    batch_size, channel, in_h, in_w = input.shape
    fh = out_h * 1.0 / in_h
    fw = out_w * 1.0 / in_w

    out = np.zeros((batch_size, channel, out_h, out_w))
    for oh in range(out_h):
        h0 = int(math.floor((oh + 0.5) / fh - 0.5))
        h1 = int(math.ceil((oh + 0.5) / fh - 0.5))
        Wh = (oh + 0.5) / fh - 0.5 - h0
        for ow in range(out_w):
            w0 = int(math.floor((ow + 0.5) / fw - 0.5))
            w1 = int(math.ceil((ow + 0.5) / fw - 0.5))
            Ww = (ow + 0.5) / fw - 0.5 - w0
            h0 = min(h0, in_h - 1)
            h1 = min(h1, in_h - 1)  
            w0 = min(w0, in_w - 1)
            w1 = min(w1, in_w - 1)
            input_h0_w0 = input[:, :, max(0, h0), max(0, w0)]
            input_h1_w0 = input[:, :, max(0, h1), max(0, w0)]
            input_h0_w1 = input[:, :, max(0, h0), max(0, w1)]
            input_h1_w1 = input[:, :, max(0, h1), max(0, w1)]
            out[:, :, oh,
                ow] = input_h0_w0 * Wh * Ww + input_h1_w0 * (
                    1 - Wh) * Ww + input_h0_w1 * Wh * (
                        1 - Ww) + input_h1_w1 * (1 - Wh) * (1 - Ww)

```

**Comment by dzarukin** [comment]
Hi @lidanqing-intel, thanks for the question.

Resampling is not a straightforward operation and has tons of edge cases.

1) The formula you refer is about boundary conditions and I don't see how it's different from the one you are using.

2) Regarding rest computations, we moved away from dividing on scale factor due to correctness issues. E.g. `h0 = int(math.floor((oh + 0.5) / fh - 0.5))` we validate as `h0 = int(math.floor((oh + 0.5) * in_h / out_h - 0.5))`. This gives the best accuracy from our testing of multiple sizes for forward and backward. Rest differences may be double checked with benchdnn: https://github.com/oneapi-src/oneDNN/blob/master/tests/benchdnn/resampling/ref_resampling.cpp. This is the way we ensure operation correctness.

3) To proceed further we will need the DNNL_VERBOSE output for failed cases (at least one or two) and the output of get versus expected.

Thanks.

**Comment by lidanqing-intel** [comment]
About point 1, you said the formula in API is no different than the one I m using? But they are different.

using API documentation: src(n, ic, ih, iw)=0 when ih =-1, iw = -1
according to formula I m using: src(n, ic, ih, iw)= src(n, ic, 0, 0) when ih = -1, iw = -1

One is set the result directly to 0, another is set the result to element at index = 0

**Comment by dzarukin** [comment]
Hi @lidanqing-intel, you convinced me something is wrong with these formulae. It's definitely not a `0` but `src(n, ic, 0, iw)` for ih < 0. But deeper look makes actually these formulae senseless since they don't represent any connection with the reality from math point of view. We will try to make math statements more precise but one should read these assumptions that border values are remain the same.
Thanks for your input on this one. If there are more things to clarify, feel free to ask them, please. Thanks.

**Comment by lidanqing-intel** [comment]
@dzarukin  I found another place in the doc that is not aligned with the reference code. The formula in the doc 
![image](https://user-images.githubusercontent.com/47151829/102000218-697e6300-3ce5-11eb-8ff3-d75c536e3417.png)
should be changed to 
```
dst[n,c,oh,ow] = src(n,c,ih0,iw0)* (1 - Wh) * (1 - Ww) + 
                 src(n,c,ih1,iw0)* Wh * (1 - Ww) + 
                 src(n,c,ih0,iw1)* (1 - Wh) * Ww +
                 src(n,c,ih1,iw1)* Wh * Ww
```
After using second formula, all my unit tests passed! Thank you for pointing me the reference code link ~

**Comment by vpirogov** [comment]
Fixed with 58e5ee23ffba463ae2f4584c2cdc49b584993ce2. Documentation is updated [online](https://oneapi-src.github.io/oneDNN/dev_guide_resampling.html).

Thank you for the report!
